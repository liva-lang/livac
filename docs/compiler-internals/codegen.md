# Code Generation

The code generator emits Rust source code from the IR, handling concurrency transformations, error handling, and runtime integration.

## Location

**File**: `src/codegen.rs` (4,683 lines)

## Overview

The code generator:
1. **Emits Rust Code**: Converts IR → valid Rust syntax
2. **Generates `liva_rt` Module**: Runtime helpers for concurrency and errors
3. **Handles Async/Par**: Transforms concurrency policies into tokio/rayon code
4. **Manages Error Binding**: Implements fallibility system
5. **Tracks Awaitable Tasks**: Lazy await/join optimization

## Architecture

### CodeGenerator

```rust
pub struct CodeGenerator {
    output: String,
    indent_level: usize,
    ctx: DesugarContext,
    in_method: bool,
    in_fallible_function: bool,
    pending_tasks: HashMap<String, TaskInfo>,
    error_binding_vars: HashSet<String>,
    fallible_functions: HashSet<String>,
    // ... other tracking fields
}
```

## Key Responsibilities

### 1. Runtime Module Generation

Generates `liva_rt` module with:

```rust
mod liva_rt {
    use std::future::Future;
    use tokio::task::JoinHandle;
    
    pub struct Error {
        pub message: String,
    }
    
    pub fn spawn_async<F, T>(future: F) -> JoinHandle<T> { ... }
    pub fn fire_async<F>(future: F) { ... }
    pub fn spawn_parallel<F, T>(f: F) -> JoinHandle<T> { ... }
    pub fn fire_parallel<F>(f: F) { ... }
    
    // Display formatting helpers
    pub fn display_array<T: std::fmt::Debug>(arr: &[T]) -> String { ... }
    // ... more helpers
}
```

### 2. Function Generation

**Non-Async, Non-Fallible**:
```rust
fn add(a: i32, b: i32) -> i32 {
    a + b
}
```

**Async**:
```rust
async fn fetchUser(id: i32) -> String {
    let response = reqwest::get(...).await;
    response.text().await.unwrap()
}
```

**Fallible**:
```rust
fn divide(a: i32, b: i32) -> Result<i32, liva_rt::Error> {
    if b == 0 {
        return Err(liva_rt::Error::from("Division by zero"));
    }
    Ok(a / b)
}
```

**Async + Fallible**:
```rust
async fn processData(url: String) -> Result<String, liva_rt::Error> {
    let data = fetch_from_api(&url).await;
    if data.is_empty() {
        return Err(liva_rt::Error::from("No data"));
    }
    Ok(data)
}
```

### 3. Concurrency Transformations

**Async Call (Immediate Await)**:
```liva
let user = async fetchUser(1)
```
↓
```rust
let user = fetchUser(1).await;
```

**Task Async (Deferred Await)**:
```liva
let userTask = task async fetchUser(1)
let user = await userTask
```
↓
```rust
let userTask = liva_rt::spawn_async(async move { fetchUser(1).await });
let user = userTask.await.unwrap();
```

**Parallel Call**:
```liva
let result = par heavyComputation(100)
```
↓
```rust
let result = { heavyComputation(100) };  // Eager evaluation
```

**Task Par**:
```liva
let task = task par compute(100)
let result = await task
```
↓
```rust
let task = liva_rt::spawn_parallel(move || compute(100));
let result = task.await.unwrap();
```

**Fire Async**:
```liva
fire async logEvent("message")
```
↓
```rust
liva_rt::fire_async(async move { logEvent("message").await; });
```

**Fire Par**:
```liva
fire par backgroundCleanup()
```
↓
```rust
liva_rt::fire_parallel(move || { backgroundCleanup(); });
```

### 4. Error Binding

**Basic Error Binding**:
```liva
let result, err = divide(10, 2)
```
↓
```rust
let (result, err) = match divide(10, 2) {
    Ok(val) => (val, "".to_string()),
    Err(e) => (Default::default(), e.to_string()),
};
```

**Error Binding with Async**:
```liva
let result, err = async divide(10, 2)
```
↓
```rust
let (result, err) = match divide(10, 2).await {
    Ok(val) => (val, "".to_string()),
    Err(e) => (Default::default(), e.to_string()),
};
```

**Error Binding with Task**:
```liva
let taskHandle = task async divide(10, 2)
let result, err = await taskHandle
```
↓
```rust
let taskHandle = liva_rt::spawn_async(async move { divide(10, 2).await });
let (result, err) = match taskHandle.await.unwrap() {
    Ok(val) => (val, "".to_string()),
    Err(e) => (Default::default(), e.to_string()),
};
```

### 5. Data-Parallel Loops

**Sequential**:
```liva
for item in items {
    process(item)
}
```
↓
```rust
for item in items {
    process(item);
}
```

**Parallel (Rayon)**:
```liva
for par item in items with threads 4 {
    process(item)
}
```
↓
```rust
items.into_par_iter()
    .with_max_threads(4)
    .for_each(|item| {
        process(item);
    });
```

**Vectorized (SIMD)**:
```liva
for vec value in values with simdWidth 4 {
    compute(value)
}
```
↓
```rust
// SIMD generation (simplified)
for chunk in values.chunks_exact(4) {
    compute(chunk[0]);
    compute(chunk[1]);
    compute(chunk[2]);
    compute(chunk[3]);
}
```

**Parallel + Vec**:
```liva
for parvec value in values with simdWidth 4 ordered {
    process(value)
}
```
↓
```rust
values.into_par_iter()
    .with_simd_width(4)
    .ordered()
    .for_each(|value| {
        process(value);
    });
```

### 6. String Templates

```liva
let name = "Alice"
let greeting = $"Hello, {name}!"
```
↓
```rust
let name = "Alice";
let greeting = format!("Hello, {}!", name);
```

Complex templates:
```liva
let msg = $"User {user.name} (age {user.age}) at {timestamp}"
```
↓
```rust
let msg = format!("User {} (age {}) at {}", user.name, user.age, timestamp);
```

### 7. Classes

```liva
Person {
    constructor(name: string, age: number) {
        this.name = name
        this.age = age
    }
    
    name: string
    age: number
    
    greet() => $"Hello, I'm {this.name}"
}
```
↓
```rust
#[derive(Debug, Clone)]
struct Person {
    name: String,
    age: i32,
}

impl Person {
    fn new(name: String, age: i32) -> Self {
        Self { name, age }
    }
    
    fn greet(&self) -> String {
        format!("Hello, I'm {}", self.name)
    }
}
```

## Main Entry Point

```rust
pub fn generate(module: ir::Module, ctx: DesugarContext) -> Result<String>
```

Returns complete Rust source code ready for compilation.

## Code Organization

### String Building

```rust
fn writeln(&mut self, s: &str)  // Write line with indentation
fn indent(&mut self)            // Increase indent
fn dedent(&mut self)            // Decrease indent
```

### Generation Methods

```rust
fn generate_function(&mut self, func: &ir::Function)
fn generate_statement(&mut self, stmt: &ir::Stmt)
fn generate_expression(&mut self, expr: &ir::Expr)
fn generate_for_loop(&mut self, for_stmt: &ir::For)
fn generate_call(&mut self, call: &ir::CallExpr)
```

## Error Codes

| Code | Description |
|------|-------------|
| **E5001** | Unsupported IR construct |
| **E5002** | Invalid concurrency combination |
| **E5003** | Codegen internal error |

## Output Format

Generated Rust code includes:

1. **Module header**: `use` statements for dependencies
2. **`liva_rt` module**: Runtime helpers
3. **Type definitions**: Structs, enums
4. **Functions**: Public and private
5. **Main function**: Entry point (if present)

## Summary

- **4,683 Lines**: Most complex compiler phase
- **IR → Rust**: Direct code emission
- **Runtime Integration**: Generates `liva_rt` module
- **Concurrency Transformation**: Handles async/par/task/fire
- **Error Binding**: Implements fallibility system
- **Data-Parallel**: Rayon and SIMD code generation

**Next**: [Desugaring →](desugaring.md)

**See Also**:
- [Runtime Module](runtime.md)
- [IR](ir.md)
- [Concurrency](../language-reference/concurrency.md)
