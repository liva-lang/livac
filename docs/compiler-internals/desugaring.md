# Desugaring

The desugaring pass collects metadata about the program and prepares it for code generation.

## Location

**File**: `src/desugaring.rs` (260 lines)

## Overview

Desugaring is a **pre-codegen analysis pass** that:
1. **Collects Dependencies**: Tracks `use rust` declarations
2. **Detects Concurrency**: Determines if tokio/rayon needed
3. **Builds Context**: Creates `DesugarContext` for codegen

## DesugarContext

```rust
pub struct DesugarContext {
    pub rust_crates: Vec<(String, Option<String>)>,  // (crate, alias)
    pub has_async: bool,
    pub has_parallel: bool,
}
```

This context tells codegen what runtime dependencies to include.

## Main Function

```rust
pub fn desugar(program: Program) -> Result<DesugarContext>
```

**Returns**: Metadata about program's runtime needs.

## Process

### 1. Collect `use rust` Declarations

```liva
use rust reqwest
use rust serde as json
```

Collected as:
```rust
rust_crates: [
    ("reqwest".to_string(), None),
    ("serde".to_string(), Some("json".to_string())),
]
```

### 2. Detect Async Usage

Checks for:
- `async` calls: `async fetchData()`
- Async functions: Functions marked `is_async_inferred`
- Async policies: `ExecPolicy::Async`, `TaskAsync`, `FireAsync`

If found:
```rust
ctx.has_async = true
rust_crates.push(("tokio".to_string(), None))  // Auto-add tokio
```

### 3. Detect Parallel Usage

Checks for:
- `par` calls: `par compute()`
- Parallel loops: `for par x in items`
- Parallel policies: `ExecPolicy::Par`, `TaskPar`, `FirePar`
- Data-parallel: `DataParallelPolicy::Par`, `Vec`, `ParVec`

If found:
```rust
ctx.has_parallel = true
// Rayon included automatically in liva_rt
```

## Example

**Input Program**:
```liva
use rust reqwest

fetchUser(id: number): string {
  let response = async httpGet($"/users/{id}")
  return response.body
}

processData(items) {
  for par item in items with threads 4 {
    compute(item)
  }
}

main() {
  let user = fetchUser(1)
  processData([1, 2, 3, 4])
}
```

**Output Context**:
```rust
DesugarContext {
    rust_crates: [
        ("reqwest", None),
        ("tokio", None),  // Auto-added
    ],
    has_async: true,     // fetchUser has async call
    has_parallel: true,  // processData has par loop
}
```

## Concurrency Detection

### Traversal Functions

```rust
fn check_concurrency(item: &TopLevel, ctx: &mut DesugarContext)
fn check_block_concurrency(body: &IfBody, ctx: &mut DesugarContext)
fn check_stmt_concurrency(stmt: &Stmt, ctx: &mut DesugarContext)
fn check_expr_concurrency(expr: &Expr, ctx: &mut DesugarContext)
```

**Recursively walks AST** to find concurrency usage.

### Async Detection

```rust
match call.exec_policy {
    ExecPolicy::Async | ExecPolicy::TaskAsync | ExecPolicy::FireAsync => {
        ctx.has_async = true;
    }
    _ => {}
}
```

### Parallel Detection

```rust
match for_stmt.policy {
    DataParallelPolicy::Par | DataParallelPolicy::Vec | DataParallelPolicy::ParVec => {
        ctx.has_parallel = true;
    }
    _ => {}
}
```

## Codegen Integration

Codegen uses `DesugarContext` to:

1. **Include Dependencies**:
   ```rust
   use tokio;
   use rayon::prelude::*;
   ```

2. **Generate Runtime Module**:
   ```rust
   mod liva_rt {
       // Include async helpers if ctx.has_async
       // Include parallel helpers if ctx.has_parallel
   }
   ```

3. **Wrap Main**:
   ```rust
   #[tokio::main]  // If ctx.has_async
   async fn main() {
       // ...
   }
   ```

## Why "Desugaring"?

The name is **legacy** from earlier designs where this pass did more transformation. Now it's primarily **metadata collection**.

**Better name**: `MetadataCollector` or `ContextBuilder`

## Summary

- **260 Lines**: Lightweight analysis pass
- **Metadata Collection**: Determines runtime needs
- **Dependency Tracking**: Manages `use rust` crates
- **Concurrency Detection**: Identifies async/parallel usage
- **Codegen Preparation**: Builds context for code generation

**Next**: [Runtime Module â†’](runtime.md)

**See Also**:
- [Code Generation](codegen.md)
- [Architecture](architecture.md)
