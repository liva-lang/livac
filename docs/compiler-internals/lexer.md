# Lexer (Tokenization)

The lexer converts source code into tokens using the **Logos** crate for high-performance lexical analysis.

## Location

**File**: `src/lexer.rs`

## Overview

The lexer performs:
1. **Tokenization**: Converts character stream to tokens
2. **Comment Stripping**: Removes single-line (`//`) and multi-line (`/* */`) comments
3. **Whitespace Handling**: Skips whitespace automatically
4. **Span Tracking**: Records source locations for error reporting

## Token Types

### Keywords (59 tokens)

**Control Flow**:
- `let`, `const`, `if`, `else`, `while`, `for`, `in`, `switch`, `case`, `default`
- `return`, `throw`, `try`, `catch`

**Concurrency**:
- `async`, `parallel`, `par`, `task`, `fire`, `await`
- `move`, `seq`, `vec`, `parvec`, `with`

**Data-Parallel Policies**:
- `ordered`, `chunk`, `threads`, `simdWidth`
- `prefetch`, `reduction`, `schedule`, `detect`
- `auto`, `safe`, `fast`, `static`, `dynamic`

**Error Handling**:
- `fail`

**Types**:
- `number`, `float`, `bool`, `char`, `string`, `bytes`
- `type`, `import`, `use`, `rust`, `test`

**Literals**:
- `true`, `false`

**Operators** (word-based):
- `and`, `or`, `not`, `as`

### Operators

**Arithmetic**: `+`, `-`, `*`, `/`, `%`

**Comparison**: `==`, `!=`, `<`, `<=`, `>`, `>=`

**Logical**: `&&`, `||`, `!`

**Other**: `=`, `.`, `..`, `=>`, `?`, `:`

### Delimiters

`(`, `)`, `{`, `}`, `[`, `]`, `,`, `;`

### Literals

```rust
IntLiteral(i64)         // 42, 100, 1_000_000
FloatLiteral(f64)       // 3.14, 1.5, 0.001
StringLiteral(String)   // "hello", "world"
CharLiteral(char)       // 'a', 'Z', '\n'
StringTemplate(String)  // $"Hello {name}"
```

### Identifiers

```rust
Ident(String)           // public: myVar, calculate
ProtectedIdent(String)  // protected: _helper
PrivateIdent(String)    // private: __secret
```

Visibility is determined by leading underscores:
- No prefix: Public
- `_`: Protected
- `__`: Private

## Implementation

### Using Logos

```rust
#[derive(Logos, Debug, Clone, PartialEq)]
#[logos(skip r"[ \t\r\n\f]+")]         // Skip whitespace
#[logos(skip r"//[^\n]*")]             // Skip single-line comments
#[logos(skip r"/\*([^*]|\*[^/])*\*/")]  // Skip multi-line comments
pub enum Token {
    #[token("let")]
    Let,
    
    #[token("async")]
    Async,
    
    #[regex(r"[0-9][0-9_]*", parse_int)]
    IntLiteral(i64),
    
    // ... more tokens
}
```

### Key Features

1. **Automatic Whitespace Skipping**: Using `#[logos(skip)]`
2. **Regex Patterns**: For numbers, strings, identifiers
3. **Underscore Separators**: `1_000_000` → `1000000`
4. **String Templates**: `$"Hello {name}"` tokenized separately

### Integer Parsing

```rust
#[regex(r"[0-9][0-9_]*", |lex| {
    lex.slice().replace('_', "").parse::<i64>().ok()
})]
IntLiteral(i64),
```

Supports underscores for readability: `1_000_000`

### Float Parsing

```rust
#[regex(r"[0-9][0-9_]*\.[0-9][0-9_]*", |lex| {
    lex.slice().replace('_', "").parse::<f64>().ok()
})]
FloatLiteral(f64),
```

### String Literals

```rust
#[regex(r#""([^"\\]|\\.)*""#, |lex| {
    let s = lex.slice();
    let content = &s[1..s.len()-1];  // Strip quotes
    Some(content.to_string())
})]
StringLiteral(String),
```

### String Templates

```rust
#[regex(r#"\$"([^"\\]|\\.)*""#, |lex| {
    let s = lex.slice();
    let content = &s[2..s.len()-1];  // Strip $" and "
    Some(content.to_string())
})]
StringTemplate(String),
```

## TokenWithSpan

```rust
pub struct TokenWithSpan {
    pub token: Token,
    pub span: Span,
}
```

Tracks source location for error reporting:
- `span.start`: Start byte offset
- `span.end`: End byte offset
- `span.start_position(map)`: (line, column)

## Public API

```rust
// Tokenize source code
pub fn tokenize(source: &str, file_name: &str) -> Result<Vec<TokenWithSpan>>
```

Returns `Vec<TokenWithSpan>` or `CompilerError` on invalid tokens.

## Error Handling

Lexer errors include:
- **Invalid tokens**: Unrecognized characters
- **Malformed literals**: Unclosed strings, invalid numbers
- **Location info**: Line, column, source snippet

## Summary

- **50+ Keywords**: Including all concurrency and data-parallel policy tokens
- **Logos-based**: High-performance lexing
- **Span Tracking**: For precise error messages
- **Comment Stripping**: Automatic via regex
- **Visibility-Aware**: Identifier prefixes determine visibility

**Next**: [Parser →](parser.md)
