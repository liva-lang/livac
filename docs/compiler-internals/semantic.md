# Semantic Analysis

The semantic analyzer validates type correctness, infers async functions, detects fallibility, and performs program-wide validation.

## Location

**File**: `src/semantic.rs` (2,067 lines)

## Overview

The semantic analyzer performs **three passes**:

1. **Definition Collection**: Gather types, functions, signatures
2. **Async Inference**: Detect which functions are async (transitive)
3. **Validation**: Type checking, fallibility checks, awaitable validation

## Key Responsibilities

### 1. Async Inference

**Automatically detects** if a function is async:

```rust
// Compiler infers this is async
fetchUser(id: number): string {
  let data = async httpGet($"/users/{id}")  // Contains async call
  return data.body
}
```

**Fixed-point iteration**: Repeats until no new async functions found.

### 2. Fallibility Detection

Tracks functions containing `fail`:

```rust
divide(a: number, b: number): number {
  if b == 0 fail "Division by zero"  // Detected as fallible
  return a / b
}
```

Fallible functions:
- Return `Result<T, liva_rt::Error>` in codegen
- Can use error binding: `let val, err = divide(10, 2)`

### 3. Type Checking

- **Function signatures**: Parameter types, return types
- **Variable types**: Tracked in symbol table
- **Type compatibility**: Ensures assignments are valid

### 4. Awaitable Validation

Enforces rules for `task` and `fire`:

- **Task handles** must be awaited before function returns
- **Fire calls** cannot be awaited
- **Pending tasks** tracked across scopes

### 5. Data-Parallel Policy Validation

Validates `for` loop policies and options:

```rust
// ❌ Invalid: seq cannot use simdWidth
for seq x in items with simdWidth 4 { }  // Error!

// ✅ Valid: parvec can use simdWidth
for parvec x in items with simdWidth 4 { }
```

**Rules enforced**:
- `seq`: No parallel/SIMD options
- `par`: Can use `chunk`, `threads`, `schedule`, `prefetch`, `reduction`, `detect`
- `vec`: Can use `simdWidth`, `ordered`
- `parvec`: Combines `par` and `vec` options

### 6. Error Binding Validation

Validates error binding patterns:

```rust
// ✅ Valid: Fallible function with 2 bindings
let result, err = divide(10, 2)

// ❌ Invalid: Non-fallible with 2 bindings
let result, err = add(10, 2)  // add() doesn't fail

// ✅ Valid: Can ignore error
let result, _ = divide(10, 2)
```

## Architecture

### SemanticAnalyzer

```rust
pub struct SemanticAnalyzer {
    async_functions: HashSet<String>,
    fallible_functions: HashSet<String>,
    types: HashMap<String, TypeInfo>,
    functions: HashMap<String, FunctionSignature>,
    external_modules: HashSet<String>,
    current_scope: Vec<HashMap<String, Option<TypeRef>>>,
    awaitable_scopes: Vec<HashMap<String, AwaitableInfo>>,
    source_file: String,
    source_code: String,
    source_map: Option<SourceMap>,
}
```

### TypeInfo

```rust
struct TypeInfo {
    name: String,
    fields: HashMap<String, (Visibility, TypeRef)>,
    methods: HashMap<String, (Visibility, bool)>, // (visibility, is_async)
}
```

### FunctionSignature

```rust
struct FunctionSignature {
    params: Vec<Option<TypeRef>>,
    return_type: Option<TypeRef>,
    is_async: bool,
    defaults: Vec<bool>,
}
```

## Public API

```rust
pub fn analyze(program: Program, source_file: &str, source_code: &str) 
    -> Result<Program>
```

Returns validated and annotated `Program`.

## Error Codes

| Code | Description |
|------|-------------|
| **E3001** | Async call outside async context |
| **E3002** | Unawaited task handle |
| **E3003** | Awaiting non-awaitable value |
| **E3004** | Fire call cannot be awaited |
| **E3005** | Await in data-parallel loop |
| **E3006** | Invalid policy option combination |
| **E3007** | Invalid chunk size |
| **E3008** | Invalid thread count |
| **E3009** | Invalid SIMD width |
| **E3010** | Error binding mismatch |

## Examples

### Async Inference

**Input AST**:
```liva
fetchData(url: string): string {
  let response = async httpGet(url)
  return response.body
}
```

**After Semantic Analysis**:
- `fetchData.is_async_inferred = true`
- Added to `async_functions` set

### Fallibility Detection

**Input AST**:
```liva
validateAge(age: number) {
  if age < 0 fail "Negative age"
  if age > 150 fail "Too old"
}
```

**After Semantic Analysis**:
- `validateAge.contains_fail = true`
- Added to `fallible_functions` set

### Error Binding Validation

**Valid**:
```liva
let result, err = divide(10, 2)  // ✅ fallible function, 2 bindings
```

**Invalid**:
```liva
let result, err = add(10, 2)  // ❌ E3010: Non-fallible with 2 bindings
```

## Summary

- **2,067 Lines**: Comprehensive semantic validation
- **Three Passes**: Collect → Infer → Validate
- **Async Inference**: Transitive, fixed-point algorithm
- **Fallibility Tracking**: Detects `fail` usage
- **Policy Validation**: Data-parallel correctness
- **Rich Errors**: E3xxx codes with source context

**Next**: [IR (Intermediate Representation) →](ir.md)

**See Also**:
- [Error System](error-system.md)
- [Concurrency](../language-reference/concurrency.md)
- [Error Handling](../language-reference/error-handling.md)
