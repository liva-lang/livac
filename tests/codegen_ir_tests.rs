use insta::assert_snapshot;
use livac::codegen::generate_from_ir;
use livac::desugaring::desugar;
use livac::lexer::tokenize;
use livac::lowering::lower_program;
use livac::parser::parse;
use livac::semantic::analyze;

fn compile_ir(source: &str) -> (String, String) {
    let tokens = tokenize(source).expect("tokenize");
    let program = parse(tokens, source).expect("parse");
    let analyzed = analyze(program.clone()).expect("semantic");
    let ctx = desugar(analyzed.clone()).expect("desugar");
    let module = lower_program(&analyzed);

    generate_from_ir(&module, &program, ctx).expect("codegen")
}

#[test]
fn ir_codegen_simple_function() {
    let source = r#"
        add(a: number, b: number): number = a + b

        main() {
            let result = add(2, 3)
            print(result)
        }
    "#;

    let (rust_code, _cargo) = compile_ir(source);
    assert_snapshot!("ir_simple_function", rust_code);
}

#[test]
fn ir_codegen_async_helpers() {
    let source = r#"
        fetch() {
            let response = async getData()
            return response
        }

        main() {
            let value = async fetch()
            fire async fetch()
            print(value)
        }
    "#;

    let (rust_code, _cargo) = compile_ir(source);
    assert_snapshot!("ir_async_helpers", rust_code);
}

#[test]
fn ir_codegen_parallel_helpers() {
    let source = r#"
        heavy() {
            return parallel compute()
        }

        main() {
            let value = parallel heavy()
            fire parallel heavy()
            print(value)
        }
    "#;

    let (rust_code, _cargo) = compile_ir(source);
    assert_snapshot!("ir_parallel_helpers", rust_code);
}

#[test]
fn ir_codegen_try_catch_and_switch() {
    let source = r#"
        main() {
            try {
                let flag = false
                if flag == false {
                    throw "error"
                }
                print("done")
            } catch (err) {
                print(err)
            }

            switch 200 {
                case 200:
                    print("success")
                case 500:
                    print("server")
                default:
                    print("other")
            }
        }
    "#;

    let (rust_code, _cargo) = compile_ir(source);
    assert_snapshot!("ir_try_catch_switch", rust_code);
}

#[test]
fn ir_codegen_string_templates() {
    let source = r#"
        main() {
            let name = "Liva User"
            let numbers = [1, 2, 3]
            let users = [
                { name: "Alice" },
                { name: "Bob" }
            ]

            print($"Name: {name}")
            print($"Numbers: {numbers}")
            print($"First user: {users[0].name}\n")
        }
    "#;

    let (rust_code, _cargo) = compile_ir(source);

    assert!(
        rust_code.contains(r#"format!("Name: {}", name)"#),
        "expected simple template to use Display placeholder:\n{}",
        rust_code
    );
    assert!(
        rust_code.contains(r#"format!("Numbers: {:?}", numbers)"#),
        "expected array template to use Debug placeholder:\n{}",
        rust_code
    );
    assert!(
        rust_code.contains(r#"format!("First user: {}\n", users[0]["name"])"#),
        "expected nested access to use bracket notation with newline escape:\n{}",
        rust_code
    );
}

#[test]
fn ir_codegen_array_param_inference_and_numeric_coercion() {
    let source = r#"
        isAdult(age) => age >= 18

        calculateTotal(items) {
            let total = 0.0
            for item in items {
                total = total + item.price
            }
            return total
        }

        main() {
            let products = [
                { price: 19.99 },
                { price: 5.50 }
            ]

            let total = calculateTotal(products)
            print(isAdult(21))
            print(total)
        }
    "#;

    let (rust_code, _cargo) = compile_ir(source);

    assert!(
        rust_code.contains("fn is_adult(age: i32) -> bool"),
        "expected isAdult to infer a bool return type:\n{}",
        rust_code
    );
    assert!(
        rust_code.contains("fn calculate_total(items: Vec<serde_json::Value>)"),
        "expected calculateTotal to infer Vec<serde_json::Value> parameter:\n{}",
        rust_code
    );
    assert!(
        rust_code.contains(".as_f64().unwrap_or(0.0)"),
        "expected numeric coercion when summing JSON values:\n{}",
        rust_code
    );
    assert!(
        rust_code.contains("let mut total = 0.0;"),
        "expected float literal to remain 0.0 and variable to be mutable:\n{}",
        rust_code
    );
}
