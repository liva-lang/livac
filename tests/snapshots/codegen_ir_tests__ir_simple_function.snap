---
source: tests/codegen_ir_tests.rs
expression: rust_code
---
mod liva_rt {
    use std::future::Future;
    use tokio::task::JoinHandle;
    
    /// Runtime error type for fallible operations
    #[derive(Debug, Clone)]
    pub struct Error {
        pub message: String,
    }
    
    impl Error {
        pub fn from<S: Into<String>>(message: S) -> Self {
            Error {
                message: message.into(),
            }
        }
    }
    
    impl std::fmt::Display for Error {
        fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
            write!(f, "{}", self.message)
        }
    }
    
    impl std::error::Error for Error {}
    
    /// Spawn an async task
    pub fn spawn_async<F, T>(future: F) -> JoinHandle<T>
    where
        F: Future<Output = T> + Send + 'static,
        T: Send + 'static,
    {
        tokio::spawn(future)
    }
    /// Fire and forget async task
    pub fn fire_async<F>(future: F)
    where
        F: Future<Output = ()> + Send + 'static,
    {
        tokio::spawn(future);
    }
    /// Spawn a parallel task
    pub fn spawn_parallel<F, T>(f: F) -> JoinHandle<T>
    where
        F: FnOnce() -> T + Send + 'static,
        T: Send + 'static,
    {
        // For simplicity, just execute synchronously and wrap in JoinHandle
        // In a real implementation, this would use rayon or std::thread
        let result = f();
        tokio::spawn(async move { result })
    }
    /// Fire and forget parallel task
    pub fn fire_parallel<F>(f: F)
    where
        F: FnOnce() + Send + 'static,
    {
        // For simplicity, just spawn a thread
        std::thread::spawn(f);
    }
    
    /// String multiplication helper
    /// Supports both String*int and int*String patterns
    pub fn string_mul<L: StringOrInt, R: StringOrInt>(left: L, right: R) -> String {
        match (left.as_string_or_int(), right.as_string_or_int()) {
            (StringOrIntValue::String(s), StringOrIntValue::Int(n)) => {
                if n <= 0 { String::new() } else { s.repeat(n as usize) }
            }
            (StringOrIntValue::Int(n), StringOrIntValue::String(s)) => {
                if n <= 0 { String::new() } else { s.repeat(n as usize) }
            }
            (StringOrIntValue::Int(a), StringOrIntValue::Int(b)) => {
                (a * b).to_string()
            }
            (StringOrIntValue::String(_), StringOrIntValue::String(_)) => {
                panic!("Cannot multiply two strings")
            }
        }
    }
    
    pub enum StringOrIntValue {
        String(String),
        Int(i64),
    }
    
    pub trait StringOrInt {
        fn as_string_or_int(self) -> StringOrIntValue;
    }
    
    impl StringOrInt for String {
        fn as_string_or_int(self) -> StringOrIntValue { StringOrIntValue::String(self) }
    }
    impl StringOrInt for &str {
        fn as_string_or_int(self) -> StringOrIntValue { StringOrIntValue::String(self.to_string()) }
    }
    impl StringOrInt for i32 {
        fn as_string_or_int(self) -> StringOrIntValue { StringOrIntValue::Int(self as i64) }
    }
    impl StringOrInt for i64 {
        fn as_string_or_int(self) -> StringOrIntValue { StringOrIntValue::Int(self) }
    }
    impl StringOrInt for f64 {
        fn as_string_or_int(self) -> StringOrIntValue { StringOrIntValue::Int(self as i64) }
    }
    impl StringOrInt for usize {
        fn as_string_or_int(self) -> StringOrIntValue { StringOrIntValue::Int(self as i64) }
    }
}

fn add(a: i32, b: i32) -> i32 {
    a + b
}

fn main() {
    let mut result = add(2, 3);
    println!("{}", result);
}
